src/liballoc/collections/vec_deque.rs

```diff
@@ -2170,12 +2170,28 @@ impl<'a, T> Iterator for Iter<'a, T> {
         back.iter().fold(accum, &mut f)
     }
 
-    fn try_fold<B, F, R>(&mut self, init: B, mut f: F) -> R where
-        Self: Sized, F: FnMut(B, Self::Item) -> R, R: Try<Ok=B>
+    fn try_fold<B, F, R>(&mut self, init: B, mut f: F) -> R
+    where
+        Self: Sized,
+        F: FnMut(B, Self::Item) -> R,
+        R: Try<Ok = B>,
     {
-        let (front, back) = RingSlices::ring_slices(self.ring, self.head, self.tail);
-        let accum = front.iter().try_fold(init, &mut f)?;
-        back.iter().try_fold(accum, &mut f)
+        let (mut iter, final_res);
+        if self.tail <= self.head {
+            // single slice self.ring[self.tail..self.head]
+            iter = self.ring[self.tail..self.head].iter();
+            final_res = iter.try_fold(init, &mut f);
+        } else {
+            // two slices: self.ring[self.tail..], self.ring[..self.head]
+            let (front, back) = self.ring.split_at(self.tail);
+            let mut back_iter = back.iter();
+            let res = back_iter.try_fold(init, &mut f);
+            self.tail = self.ring.len() - back_iter.len();
```

Comment:
`next` will checked read from `tail`, so I wanted to see if I could convince myself that a back-iter of zero length wouldn't ever be seen here, but I don't think that's guaranteed.  Take this test:

```rust
    use std::collections::VecDeque;

    let mut v = VecDeque::with_capacity(8);
    v.push_back(7);
    v.push_back(8);
    v.push_back(9);
    v.push_front(2);
    v.push_front(1);
    assert_eq!(v.as_slices(), (&[1, 2][..], &[7, 8, 9][..]));
    let mut it = v.iter();
    it.find(|&&x| x == 2);
    assert_eq!(it.next(), Some(&7));
```

That sets up the storage as 
```
7 8 9 _ _ _ 1 2
    h       t
```
then the `find` call returns an break from the first try_fold, but after consuming all the items.  So `tail` gets set to `ring.len()`, and then the function returns early from the `?` on line 2191.  Then the call to `next` reads uninitialized memory.

Does that make sense, or is there something I missed?

Reply:
You're right. I need a wrapping operation on the `fold` (though not on the `rfold`, because in that case the iterator is already placed at the correct element, right?).

Reply:
I think so, since neither slice can be as big as the underlying capacity.